plugins/builds/index.js

```diff
@@ -785,18 +820,86 @@ const buildsPlugin = {
             const { pipeline, job, build } = config;
             const { eventFactory, pipelineFactory, buildFactory, jobFactory } = app;
             const event = await eventFactory.get({ id: build.eventId });
+            const stageBuilds = await event.getStageBuilds();
+            const stageWorkflowGraphs = {};
+
+            stageBuilds.forEach(stageBuild => {
+                stageWorkflowGraphs[stageBuild.stageId] = stageBuild.workflowGraph;
+            });
+
+            const flattenedWorkflowGraph = workflowParser.getFlattenedWorkflow(
+                event.workflowGraph,
+                stageWorkflowGraphs
+            );
             const current = {
                 pipeline,
                 job,
                 build,
-                event
+                event,
+                stageBuilds,
+                flattenedWorkflowGraph
             };
-
-            const nextJobsTrigger = workflowParser.getNextJobs(current.event.workflowGraph, {
+            let nextJobsTrigger = workflowParser.getNextJobs(current.flattenedWorkflowGraph, {
                 trigger: current.job.name,
                 chainPR: pipeline.chainPR
             });
 
+            const currentNode = current.flattenedWorkflowGraph.nodes.find(n => n.name === current.job.name);
+
+            // Figure out if need to run stage teardown
+            // If build is in stage
+            if (currentNode && currentNode.stageName) {
+                const currentStageBuild = stageBuilds.filter(sb => sb.stageName === currentNode.stageName)[0];
+
+                // Set stageBuild status to running if setup
+                if (currentNode.name === `stage@${currentNode.stageName}:setup`) {
+                    currentStageBuild.status = 'RUNNING';
+                    await currentStageBuild.update();
+                    // Set stageBuild status if completed
+                } else {
+                    //  else get all stage builds in workflow, compare number of stageBuilds to stage builds with status
+                    // Get job names from stage workflow
+                    const workflowStageBuilds = [];
+
+                    current.flattenedWorkflowGraph.nodes.forEach(n => {
+                        if (n.stageName === currentNode.stageName) {
+                            workflowStageBuilds.push(n.name);
+                        }
+                    });
+
+                    // get all builds in current stage
+                    const buildParentBuilds = current.build.parentBuilds;
```

Comment:
Relying on `parentBuilds` may not work for all scenarios.
For example: Below stage has 6 jobs on two separate paths running in parallel
![image](https://github.com/screwdriver-cd/screwdriver/assets/2124590/abae7b6d-f5cb-41a5-8f03-a3d6bbfa01e4)
Parent builds for `prod-blue-certify` will only include `prod-blue-deploy` and `prod-blue-test`.
So the `teardown` will not be triggered when the all the jobs in the stage are successfully completed.

Reply:
Another scenario could be a stage with only one job.